go/libraries/doltcore/schema/typeinfo/varstring.go

```diff
@@ -0,0 +1,302 @@
+// Copyright 2020 Liquidata, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package typeinfo
+
+import (
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+
+	"github.com/src-d/go-mysql-server/sql"
+	"vitess.io/vitess/go/sqltypes"
+
+	"github.com/liquidata-inc/dolt/go/store/types"
+)
+
+const (
+	varStringTypeParam_Collate              = "collate"
+	varStringTypeParam_Length               = "length"
+	varStringTypeParam_RemoveTrailingSpaces = "rts"
+	varStringTypeParam_IsSqlText            = "text"
+)
+
+type varStringImpl struct {
+	Collation            sql.Collation
+	MaxLength            int64
+	RemoveTrailingSpaces bool
+	IsSqlText            bool // When converting to a SQL type, this ensures a distinction between TEXT and VARCHAR
+}
+
+var _ TypeInfo = (*varStringImpl)(nil)
+var StringDefaultType TypeInfo = &varStringImpl{
+	sql.Collation_Default,
+	math.MaxUint32,
+	false,
+	true,
+}
+
+func CreateVarStringTypeFromParams(params map[string]string) (TypeInfo, error) {
+	ti := &varStringImpl{}
+	var err error
+	if collationStr, ok := params[varStringTypeParam_Collate]; ok {
+		ti.Collation, err = sql.ParseCollation(nil, &collationStr, false)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		return nil, fmt.Errorf(`create varstring type info is missing param "%v"`, varStringTypeParam_Collate)
+	}
+	if maxLengthStr, ok := params[varStringTypeParam_Length]; ok {
+		ti.MaxLength, err = strconv.ParseInt(maxLengthStr, 10, 64)
+	} else {
+		return nil, fmt.Errorf(`create varstring type info is missing param "%v"`, varStringTypeParam_Length)
+	}
+	if _, ok := params[varStringTypeParam_RemoveTrailingSpaces]; ok {
+		ti.RemoveTrailingSpaces = true
+	}
+	if _, ok := params[varStringTypeParam_IsSqlText]; ok {
+		ti.IsSqlText = true
+	}
+	return ti, nil
+}
+
+// ConvertNomsValueToValue implements TypeInfo interface.
+func (ti *varStringImpl) ConvertNomsValueToValue(v types.Value) (interface{}, error) {
+	if val, ok := v.(types.String); ok {
+		if ti.RemoveTrailingSpaces {
+			return strings.TrimRight(string(val), " "), nil
```

Comment:
From the MySQL documentation:
```
If a given value is stored into the CHAR(4) and VARCHAR(4) columns, the values retrieved from the columns are not always the same because trailing spaces are removed from CHAR columns upon retrieval.
```